### ISSUE TITLE
TinyTDS::Error Unclosed quotation mark after the character string... We can't save into image data type any longer. 

### ISSUE THREAD

**Issue by alanruth** [rails-sqlserver/activerecord-sqlserver-adapter#405]
We are working to upgrade our application to a more current version of Ruby & Rails. Our app integrates with a legacy database (SQL Server 2008 R2) that has a table with a column of image data type (we are unable to change this column to varbinary(max)). Previously we were able to save a binary into the image column. However now we are getting conversion errors.

We are working to upgrade to the following (among others):

* Rails 4.2.1
* ActiveRecord_SQLServer_Adapter (4.2.4)
* tiny_tds (0.6.3.rc1)
* freeTDS (v0.91.112)

As part of this upgrade, we had to add the following configuration as there are triggers on the legacy table being inserted: 

``` 
ActiveRecord::ConnectionAdapters::SQLServerAdapter.use_output_inserted = true
```
Removing this configuration produces the following error:  
```
TinyTds::Error: The target table 'doc' of the DML statement cannot have any enabled triggers if the statement contains an OUTPUT clause without INTO clause.
```

Now when we attempt to save into the image column, we get errors similar to:

```
TinyTds::Error: Unclosed quotation mark after the character string
```

Researching various issues within tiny_tds & activerecord_sqlserver_adapter, we decided to create a second table that matched the first but change the data type from image to varbinary(max). We can save a binary into the column. [we don't control the legacy database and do know that image data type is being deprecated in future sql server releases]. 

The code producing the error is in a background job where we grab images from s3, store them locally and then push the image into the database. Again, we don't control the legacy database and thus can't change the data type (or confront the issue of why we are storing the image in the db in the first place).

```
...
@d = Doc.new
...
open("#{Rails.root}/cache/pictures/image.png", "wb") do |file|
    file << open(r.image.url).read
end

@d.document = File.binread("#{Rails.root}/cache/pictures/image.png")
@d.save!
```

We are trying to figure out how best to determine a fix. Anyone have any ideas on how to proceed? None of us have deep knowledge of tiny_tds or activerecord_sqlserver_adapter.

**Comment by sixfeetover** [comment]
While I'm not sure about your particular error, I do know that version 4.2 of the adapter targets SQL Server 2014+. From the readme:

> The SQL Server adapter for ActiveRecord v4.2 using SQL Server 2012 or higher. If you need the adapter for SQL Server 2008 or 2005, you are still in the right spot. Just install the latest 3.2.x to 4.1.x version of the adapter. 

That means the highest you can go with SQL Server 2008 R2 is Rails 4.1 with version 4.1 of the adapter.

We are also on 2008 R2 for now, and have added [this snippet](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/issues/339#issuecomment-49351311) to our 4.1 Rails apps until we can upgrade to 4.2.

**Comment by metaskills** [comment]
Yes, 4.2 is only for 2012+

**Comment by alanruth** [comment]
Thank you gentlemen! We will fall back to 4.1.

**Comment by redbar0n** [comment]
I just ran into this problem too. 

--- 
Our specs:

- SQL Server 2016 Service Pack 1 (SP1) (build 13.0.4001.0)
- rails (4.1.16)
- activerecord-sqlserver-adapter (4.1.7)
- tiny_tds (2.0.0)
- freetds v0.95.81

--- 
Our case

When someone tried to update the attributes of a certain record, we got:
`TinyTds::Error: Unclosed quotation mark after the character string 'Forny førerkortet '.`

--- 
Extra info

We're not specifying `SQLServerAdapter.use_output_inserted` anywhere.

> That means the highest you can go with SQL Server 2008 R2 is Rails 4.1 with version 4.1 of the adapter.

I interpret this as: you cannot go higher than rails 4.1 and 4.1 of sqlserver-adapter with SQL server 2008, because that's the highest version they support. But it doesn't say that you cannot go to SQL Server 2016. It would be strange if 2016 wasn't backwards compatible with the older adapters..

--- 

Any suggestions?

**Comment by redbar0n** [comment]
Also getting it another place.

`TinyTds::Error: Unclosed quotation mark after the character string ''.`

This could be the answer:

> The way I was able to correct this issue, was converting my string I was trying to save from ISO-8859-1 to UTF-8 by doing the following code: my_string.encode!('UTF-8'). This removed the invalid characters that were causing and issue
- daveomcd at https://stackoverflow.com/questions/41245034/how-to-handle-unclosed-quotation-mark-after-the-character-string-in-ruby

We're receiving data from app clients. And I just reproduced the issue by submitting some smileys in a text field in the apps.

**Comment by sixfeetover** [comment]
You need freetds > 1 to support emoji or other multi-byte characters. If you are unable to upgrade you can also try setting `use utf-16 = yes` (yes, 16, not 8) in freetds.conf.

**Comment by redbar0n** [comment]
@sixfeetover I now installed FreeTDS 1.00.80 which supports utf-16 by default, and ran `bundle exec gem uninstall tiny_tds` and `bundle install` again, to compile tiny_tds with the binding to the new FreeTDS version. This made the following warning disappear from the rails console:

/home/konsulent.shortcut/mullet/vendor/bundle/ruby/2.3.0/gems/tiny_tds-2.0.0/lib/tiny_tds/client.rb:53: warning: TinyTds: :use_utf16 option not supported in this version of FreeTDS.

Thus I would presume it would be able to handle multi-byte characters. But I still get:

´TinyTds::Error: Unclosed quotation mark after the character string ''.´

When storing to the DB.

**Comment by metaskills** [comment]
> But it doesn't say that you cannot go to SQL Server 2016. It would be strange if 2016 wasn't backwards compatible with the older adapters..

True, it should work but we do not actively test the really old branches. But yes, in theory, the Rails `4-1-stable` branch should work just fine with all the latest FreeTDS, SQL Server, and TinyTDS. That said, I did not do the work for that version and I fear it has some edge case bugs.

> But I still get TinyTds::Error: Unclosed quotation mark after the character string.

I would highly recommend going to the v4.2 Rails and upgrade to that version of the adapter. The new ActiveRecord type system came in 4.2 which I did a lot of good work on.

**Comment by sixfeetover** [comment]
Yeah, I would echo what @metaskills said - 4.0 and 4.1 of this adapter was a bit hairy, but 4.2 and 5.0 are solid.

My guess is if you tried using TinyTDS directly it would work. This is what I used when I was experimenting with this back in 2015, just to see if wide byte characters could survive the round trip:

```ruby
# emoji.rb
require 'tiny_tds'

client = TinyTds::Client.new # ... your config

puts client.charset
face = client.execute("select N'😝' face, N'五' jp").each
puts face, face.first['face'], face.first['jp']
```
```
# freetds.conf
[global]
dump file = /tmp/freetds.log
debug flags = 0xffff
encryption = require

#tds version = 8.0
#use utf-16 = yes
```

You could then run with the custom freetds.conf file, playing with the flags, with `FREETDS=freetds.conf ruby emoji.rb`, or without it with just `ruby emoji.rb`. You should _not_ need a custom freetds.conf with the latest tiny_tds and freetds. With the config I've given, you can watch the raw traffic by tailing `/tmp/freetds.log`.

If that works, it means it's a problem with Rails 4.1 or this adapter, and upgrading to 4.2 or 5 might solve the problem.

**Comment by redbar0n** [comment]
Thanks! I managed to get it working! 

I was previously doing `params[:notes].encode!('UTF-8')` (also tried `.force_encoding('UTF-8')`)on the string as a temporary fix to successfully deal with UTF-8 characters like `æ`, `ø`, and `å`, without having to upgrade FreeTDS.

But after I upgraded the FreeTDS version then it was actually that encode! call which preceded to give the exact same error (or was it the `.force_encoding` call? I can't quite remember). When removing that call entirely, then it worked with emojis. :-)

I guess the parameters are coming in as UTF-16 and then ruby/rails just passes it on like UTF-16 to tiny_tds to the updated FreeTDS version which accepts it. **Or**, ruby reads the params string as UTF-8 (Ruby 2.0 uses UTF-8 by default, no?) but doesn't explicitly try to transcode the unknown emoji bytes as UTF-8, so it stores/retrieves it to DB just fine. (Man, I hate encoding-issues!)

I made a reinstall_freetds.sh script, which helped me a lot (to replicate the working reinstallation procedure across multiple servers). It's somewhat specific to our case, but in case some of it might help someone else, here is the script (pretty well documented): https://gist.github.com/redbar0n/220dcf53dc0b9b025d1b8c4a83dc567d

Note: All of our other system specs remained the same (as those stated above). So you might be relieved to know tiny_tds 2.0.0 and FreeTDS 1.00.80 does work with UTF-16 emoji characters with Rails 4.1.16.
